### PR TITLE
Tiktok sigi

### LINKED ISSUE #3551: Can't download tiktok video
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.04.08** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Russia

### Description

Video link is available, but yt-dlp can't download it:
https://www.tiktok.com/@denidil6/video/7065799023130643713

### Verbose log

```shell
D:\PROGRAMS\TikTok Download>!dl https://www.tiktok.com/@denidil6/video/7065799023130643713 -vU
[debug] Command-line config: ['https://www.tiktok.com/@denidil6/video/7065799023130643713', '-vU']
[debug] Encodings: locale cp1251, fs utf-8, out utf-8, err utf-8, pref cp1251
[debug] yt-dlp version 2022.04.08 [7884ade] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg n5.0.1-3-gb655beb025-20220419 (setts), ffprobe n5.0.1-3-gb655beb025-20220419
[debug] Optional libraries: brotli, certifi, Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2022.04.08, Current version: 2022.04.08
yt-dlp is up to date (2022.04.08)
[debug] [TikTok] Extracting URL: https://www.tiktok.com/@denidil6/video/7065799023130643713
[TikTok] 7065799023130643713: Downloading video details
WARNING: [TikTok] 7065799023130643713: Video not available; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U; Retrying with feed workaround
[TikTok] 7065799023130643713: Downloading video feed
WARNING: [TikTok] 7065799023130643713: Unable to find video in feed; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U; Retrying with webpage
[TikTok] 7065799023130643713: Downloading webpage
ERROR: [TikTok] 7065799023130643713: Unable to download webpage: The read operation timed out (caused by timeout('The read operation timed out')); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp\extractor\common.py", line 641, in extract
  File "yt_dlp\extractor\tiktok.py", line 520, in _real_extract
  File "yt_dlp\extractor\common.py", line 932, in _download_webpage
  File "yt_dlp\extractor\common.py", line 800, in _download_webpage_handle
  File "yt_dlp\extractor\common.py", line 785, in _request_webpage

  File "yt_dlp\extractor\common.py", line 767, in _request_webpage
  File "yt_dlp\YoutubeDL.py", line 3601, in urlopen
  File "urllib\request.py", line 525, in open
  File "urllib\request.py", line 542, in _open
  File "urllib\request.py", line 502, in _call_chain
  File "yt_dlp\utils.py", line 1543, in https_open
  File "urllib\request.py", line 1358, in do_open
  File "http\client.py", line 1344, in getresponse
  File "http\client.py", line 307, in begin
  File "http\client.py", line 268, in _read_status
  File "socket.py", line 669, in readinto
  File "ssl.py", line 1241, in recv_into
  File "ssl.py", line 1099, in read
socket.timeout: The read operation timed out
```

### PR DESCRIPTION
**Pull request by sulyi**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

### Description of your *pull request* and other information

This adds extraction of SIGI_STATE feature from https://github.com/ytdl-org/youtube-dl/pull/30479 and implements its usage in `TikTokIE`.

### Tests

- [x] TikTok
- [x] Tiktok_1
- [x] Tiktok_2
- [ ] ~~Tiktok_3~~ (unavailable, confirmed)
- [x] Tiktok_4

### CODE REVIEW

File: yt_dlp/extractor/tiktok.py [hunk 1]
```diff
@@ -511,23 +572,20 @@ def _real_extract(self, url):
...
         except ExtractorError as e:
-            self.report_warning(f'{e}; Retrying with webpage')
+            self.report_warning(f'{e.orig_msg}; Retrying with webpage')
 
-        # If we only call once, we get a 403 when downlaoding the video.
-        self._download_webpage(url, video_id)
-        webpage = self._download_webpage(url, video_id, note='Downloading video webpage')
+        # If UA starts with `Mozilla/5.0`, but shorter than 26 character, 403 response is received
+        webpage = self._download_webpage(
+            url, video_id, headers={'User-Agent': 'Mozilla/5.0 innocent fishing boat'})
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:579]:
```suggestion
        # If UA starts with `Mozilla/5.0`, but shorter than 26 character, 403 response is received
        webpage = self._download_webpage(
            url, video_id, headers={'User-Agent': 'Mozilla/5.0 innocent fishing boat'})
```
```suggestion
        webpage = self._download_webpage(
            url, video_id, headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:99.0) Gecko/20100101 Firefox/99.0'})
```

Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:579]:
```suggestion
        # If UA starts with `Mozilla/5.0`, but shorter than 26 character, 403 response is received
        webpage = self._download_webpage(
            url, video_id, headers={'User-Agent': 'Mozilla/5.0 innocent fishing boat'})
```
```suggestion
        webpage = self._download_webpage(
            url, video_id, headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:99.0) Gecko/20100101 Firefox/99.0'})
```

File: yt_dlp/extractor/tiktok.py [hunk 2]
```diff
@@ -492,14 +547,20 @@ class TikTokIE(TikTokBaseIE):
...
         except ExtractorError as e:
-            self.report_warning(f'{e}; Retrying with feed workaround')
-            feed_list = self._call_api('feed', {'aweme_id': aweme_id}, aweme_id,
-                                       note='Downloading video feed', errnote='Unable to download video feed').get('aweme_list') or []
+            self.report_warning(f'{e.orig_msg}; Retrying with feed workaround')
+            res = self._call_api(
+                'feed', {'aweme_id': aweme_id}, aweme_id,
+                note='Downloading video feed', errnote='Unable to download video feed'
+            )
+            feed_list = res.get('aweme_list') or []
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:563]: unnecessary change

Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:563]: unnecessary change

File: yt_dlp/extractor/tiktok.py [hunk 3]
```diff
@@ -492,14 +547,20 @@ class TikTokIE(TikTokBaseIE):
 
     def _extract_aweme_app(self, aweme_id):
         try:
-            aweme_detail = self._call_api('aweme/detail', {'aweme_id': aweme_id}, aweme_id,
-                                          note='Downloading video details', errnote='Unable to download video details').get('aweme_detail')
+            res = self._call_api(
+                'aweme/detail', {'aweme_id': aweme_id}, aweme_id,
+                note='Downloading video details', errnote='Unable to download video details'
+            )
+            aweme_detail = res.get('aweme_detail')
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:554]: same

Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:554]: same

File: yt_dlp/extractor/tiktok.py [hunk 4]
```diff
@@ -141,6 +142,27 @@ def _get_subtitles(self, aweme_detail, aweme_id):
             })
         return subtitles
 
+    def _get_SIGI_STATE(self, video_id, html):
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:145]:
```suggestion
    def _get_sigi_state(self, video_id, html):
```

Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:145]:
```suggestion
    def _get_sigi_state(self, video_id, html):
```
```suggestion
    def _get_SIGI_STATE(self, video_id, html):
```

File: yt_dlp/extractor/tiktok.py [hunk 5]
```diff
@@ -141,6 +142,27 @@ def _get_subtitles(self, aweme_detail, aweme_id):
...
+        try:
+            sigi_data = get_element_by_id('SIGI_STATE', html)
+            if not sigi_data:
+                raise ExtractorError('No element with \'SIGI_STATE\' id was found')
+        except (ValueError, ExtractorError) as err:
+            if isinstance(err, ExtractorError):
+                self.write_debug(err.orig_msg)
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:152]: why?

File: yt_dlp/extractor/tiktok.py [hunk 6]
```diff
@@ -141,6 +142,27 @@ def _get_subtitles(self, aweme_detail, aweme_id):
...
+                    <script\s[^>]*?\b
+                        id\s*=\s*(?P<q>"|'|\b)
+                        sigi-persisted-data(?P=q)[^>]*>[^=]*=\s*(?P<json>{.+?})\s*(?:;[^<]+)?
+                    </script
+                ''', html, 'sigi data', default='{}', group='json')
+        try:
+            state = self._parse_json(sigi_data, video_id)
+        except TypeError as err:
+            raise ExtractorError('No SIGI data was found') from err
+        return state if isinstance(state, dict) else {}
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:164]: why are you catching and rethrowing the errors!?

Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:164]: why are you catching and rethrowing the errors!?

File: yt_dlp/extractor/tiktok.py [hunk 7]
```diff
@@ -141,6 +142,27 @@ def _get_subtitles(self, aweme_detail, aweme_id):
...
+                raise ExtractorError('No element with \'SIGI_STATE\' id was found')
+        except (ValueError, ExtractorError) as err:
+            if isinstance(err, ExtractorError):
+                self.write_debug(err.orig_msg)
+            sigi_data = self._search_regex(
+                r'''(?s)
+                    <script\s[^>]*?\b
+                        id\s*=\s*(?P<q>"|'|\b)
+                        sigi-persisted-data(?P=q)[^>]*>[^=]*=\s*(?P<json>{.+?})\s*(?:;[^<]+)?
+                    </script
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:158]: regex is wrong

Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:158]: regex is wrong

File: yt_dlp/extractor/tiktok.py [hunk 8]
```diff
@@ -511,23 +535,31 @@ def _real_extract(self, url):
...
         except ExtractorError as e:
-            self.report_warning(f'{e}; Retrying with webpage')
+            self.report_warning(f'{e.orig_msg}; Retrying with webpage')
 
-        # If we only call once, we get a 403 when downlaoding the video.
-        self._download_webpage(url, video_id)
-        webpage = self._download_webpage(url, video_id, note='Downloading video webpage')
+        # If UA starts with `Mozilla/5.0`, but shorter than 26 character, 403 response is received
+        webpage = self._download_webpage(
+            url, video_id, headers={'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:98.0) Gecko/20100101 Firefox/98.0'})
```
Comment (sulyi) [inline yt_dlp/extractor/tiktok.py:542]:
```suggestion
        # If we only call once, we get a 403 when downlaoding the video.
        webpage = self._download_webpage(url, video_id, tries=2)
```
I'm stashing this, when `random_user_agent` gets fixed.
Tho, it seams to be working with a single request atm.

File: yt_dlp/extractor/tiktok.py [hunk 9]
```diff
@@ -546,14 +546,17 @@ def _real_extract(self, url):
...
+            try:
+                sigi_data = get_element_by_id('SIGI_STATE', webpage)
+            except ValueError:
+                sigi_data = self._search_regex(
+                    r'''(?s)
+                        <script\s[^>]*?\b
+                            id\s*=\s*(?P<q>"|'|\b)
+                            SIGI_STATE(?P=q)[^>]*>[^=]*=\s*(?P<json>{.+?})\s*(?:;[^<]+)?
+                        </script
+                    ''', webpage, 'sigi data', group='json')
```
Comment (sulyi) [inline yt_dlp/extractor/tiktok.py:558]: Still, no good.

File: yt_dlp/extractor/tiktok.py [hunk 10]
```diff
@@ -511,23 +535,30 @@ def _real_extract(self, url):
...
-            video_data = traverse_obj(sigi_data, ('ItemModule', video_id), expected_type=dict)
+            try:
+                sigi_data = get_element_by_id('SIGI_STATE', webpage)
+            except ValueError:
+                sigi_data = self._search_regex(
+                    '(?s)<script\\s[^>]*?\\b'
+                    'id\\s*=\\s*(?P<q>"|\'|\\b)SIGI_STATE(?P=q)'
+                    '[^>]*>\\s*'
+                    '(?P<json>\\{.+?\\})\\s*(?:;[^<]+)?'
+                    '</script', webpage, 'sigi data', group='json')
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:557]:
```suggestion
                sigi_data = self._search_regex(
                    r'(?s)<script\s[^>]*\bid\s*=\s*(?P<q>["\'])SIGI_STATE(?P=q)[^>]*>\s*(?P<json>{.+?})\s*;',
                    webpage, 'sigi data', group='json')
```
Give an example URL where this regex is needed

Comment (sulyi) [reply yt_dlp/extractor/tiktok.py:557]: There's none among test cases. It's only possible if `webpage` contains invalid tag afak.
Should it be safeguarded by adding (_unexpected_) warning to `except` clause?

**Review by sulyi — COMMENTED** [review]
Might not be necessary, it's probably legacy behaviour of the site.

File: yt_dlp/extractor/tiktok.py [hunk 11]
```diff
@@ -547,11 +547,11 @@ def _real_extract(self, url):
...
         else:
             try:
-                sigi_data = get_element_by_id('SIGI_STATE', webpage)
+                sigi_data = get_element_by_attribute(r'(SIGI_STATE|sigi-persisted-data)', webpage, escape_value=False)
             except ValueError:
                 sigi_data = self._search_regex(
                     '(?s)<script\\s[^>]*?\\b'
-                    'id\\s*=\\s*(?P<q>"|\'|\\b)SIGI_STATE(?P=q)'
+                    'id\\s*=\\s*(?P<q>"|\'|\\b)(SIGI_STATE|sigi-persisted-data)(?P=q)'
                     '[^>]*>[^=]*?=?\\s*'
```
Comment (sulyi) [inline yt_dlp/extractor/tiktok.py]: Rejecter (previous commit included) by upstream.

File: yt_dlp/extractor/tiktok.py [hunk 12]
```diff
@@ -556,7 +590,7 @@ class TikTokUserIE(TikTokBaseIE):
             'title': '6820838815978423302',
             'thumbnail': r're:https://.+_1080x1080\.webp'
         },
-        'expected_warnings': ['Retrying']
+        'expected_warnings': ['Retrying', 'Creating a generic title instead']
```
Comment (sulyi) [inline yt_dlp/extractor/tiktok.py:593]: Is https://github.com/yt-dlp/yt-dlp/pull/3549#issuecomment-1108589347 applicable here?

File: yt_dlp/extractor/tiktok.py [hunk 13]
```diff
@@ -710,7 +744,7 @@ class TikTokEffectIE(TikTokBaseListIE):
         'info_dict': {
             'id': '1258156',
         },
-        'expected_warnings': ['Retrying']
+        'expected_warnings': ['Retrying', 'Creating a generic title instead']
```
Comment (sulyi) [inline yt_dlp/extractor/tiktok.py:747]: And here?

File: yt_dlp/extractor/tiktok.py [hunk 14]
```diff
@@ -565,7 +599,7 @@ class TikTokUserIE(TikTokBaseIE):
             'title': 'meme',
             'thumbnail': r're:https://.+_1080x1080\.webp'
         },
-        'expected_warnings': ['Retrying']
+        'expected_warnings': ['Retrying', 'Creating a generic title instead']
```
Comment (sulyi) [inline yt_dlp/extractor/tiktok.py:602]: Or here?

**Review by sulyi — COMMENTED** [review]
Testcase changes question:

**Comment by pukkandan** [comment]
> Is https://github.com/yt-dlp/yt-dlp/pull/3549#issuecomment-1108589347 applicable here?

No, because you are modifying the extractor. I only meant to explain why the tests fail and why it is not important to fix them **when not making changes to the extractor itself**

**Comment by sulyi** [comment]
> because you are modifying the extractor

I understand, yes, I am doing that to `TikTokIE`.

But the changes to testcases I'm asking about belong to `TikTokEffectIE` and `TikTokUserIE`. I didn't touch those.

https://github.com/yt-dlp/yt-dlp/pull/3624/commits/49517fa96d82f6ed756ed46f3119d2c3b9a1d3e0, https://github.com/yt-dlp/yt-dlp/pull/3624/commits/398cf7696ef6039dac74e38dbe47923853a20ef0

File: yt_dlp/extractor/tiktok.py (hunk 10 above)
Comment (pukkandan) [reply yt_dlp/extractor/tiktok.py]: > It's only possible if webpage contains invalid tag afak.

and this is known to happen? If so, provide an example URL where this code is reached. if not, I do not understand why this is needed. This is looking for the same thing as the above get_element_by_attribute. Are you suggesting that all code that uses `get_element_by_attribute` must have a fallback? That makes no sense.

File: yt_dlp/extractor/tiktok.py [hunk 15]
```diff
@@ -511,23 +535,43 @@ def _real_extract(self, url):
...
+            alnum = string.ascii_letters + string.digits
+            # text: printable not in whitespace
+            name_set = '!#$%&\'*+-.^_`|~' + alnum  # in text not in '()<>@,;:"/[]?={}\\'
+            value_set = '!#$%&\'()*+-./:<=>?@[]^_`{|}~' + alnum  # in text not in '",;\\'
+            name_length = random.randint(1, 4096 - 1)  # keep at least one byte for value
+            value_length = random.randint(1, 4096 - name_length)
+            return ''.join(random.choice(name_set) for _ in range(name_length)), \
+                   ''.join(random.choice(value_set) for _ in range(value_length))
+
+        self._set_cookie('www.tiktok.com', *get_random_cookie())
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:551]: Is all this code really needed to create a random cookie? Just creating a small random key and value from `ascii_letters` would significantly reduce this code. I assume we don't care about uniform distrinution of the generated cookies.

PS: Why is a random cookie needed anyway?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/tiktok.py [hunk 16]
```diff
@@ -485,7 +485,6 @@ class TikTokIE(TikTokBaseIE):
             'repost_count': int,
             'comment_count': int,
         },
-        'expected_warnings': ['Creating a generic title']
```
Comment (sulyi) [inline yt_dlp/extractor/tiktok.py]: This one is valid, tho, I'll readd.

Comment (pukkandan) [reply yt_dlp/extractor/tiktok.py]: No!

File: yt_dlp/extractor/tiktok.py (hunk 15 above)
Comment (sulyi) [reply yt_dlp/extractor/tiktok.py:551]: > compliant with RFC 6265

I'm not sure, but it prevent error in #3551. This is still just a workaround, but better than setting UA.
I'm planning to move this in some where else, not sure where.

File: yt_dlp/extractor/tiktok.py (hunk 16 above)
Comment (sulyi) [reply yt_dlp/extractor/tiktok.py]: Sry, I missed that.

File: yt_dlp/extractor/tiktok.py [hunk 17]
```diff
@@ -557,21 +556,10 @@ def get_random_cookie():
...
-                        <script\s[^>]*?\b
-                            id\s*=\s*(?P<q>"|'|\b)(SIGI_STATE|sigi-persisted-data)(?P=q)[^>]*>
-                            [^=]*?=?\s*(?P<json>\{.+?\})\s*(?:;[^<]+)?
-                        </script
-                    ''', webpage, 'sigi data', group='json')
-            sigi_state = self._parse_json(sigi_data, video_id)
-
-            status = traverse_obj(sigi_state, ('VideoPage', 'statusCode'), expected_type=int) or 0
-            video_data = traverse_obj(sigi_state, ('ItemModule', video_id), expected_type=dict)
+            sigi_json = get_element_by_id('SIGI_STATE', webpage)
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:559]: any reason u dropped `sigi-persisted-data`?

Comment (sulyi) [reply yt_dlp/extractor/tiktok.py:559]: More than one:
It seems to be a legacy behaviour, no longer sent.
Judging from the regex it was assigned to a js variable which would require to process it before parsing as json.

**Comment by sulyi** [comment]
Length of the cookie is limited in 4096 byte, that's includes name, value all other attribute names and values and separators (equal sign, space and semicolon). All these requirements can't be met doing it here or in `utils`.

**Comment by pukkandan** [comment]
> Length of the cookie is limited in 4096 byte, that's includes name, value all other attribute names and values and separators (equal sign, space and semicolon). All these requirements can't be met doing it here or in utils.

https://github.com/yt-dlp/yt-dlp/pull/3624#discussion_r867358459

**Comment by afterdelight** [comment]
why did you close this?
